clash-lib/src/Clash/Primitives/DSL.hs

```diff
@@ -220,21 +224,27 @@ declaration
   => Text.Text
   -- ^ block name
   -> State (BlockState backend) ()
+  -- ^ action to lay a component declaration in the appropriate place (VHDL)
+  -> State (BlockState backend) ()
   -- ^ block builder
   -> State backend Doc
   -- ^ pretty printed block
-declaration blockName s = do
+declaration blockName c s = do
   backend0 <- get
-  let initState = BlockState [] IntMap.empty backend0
-      BlockState decs _hoCalls backend1 = execState s initState
+  let initState = BlockState [] [] IntMap.empty backend0
+      (BlockState cdecs decs _hoCalls backend1) = execState (c*>s) initState
```

Comment:
It's probably time to make an `emptyBlockState` top-level binder. Initialize it using its field names for clarity. Probably also wise to use `NamedFieldPuns` when deconstructing.